content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,339 @@ Like a segment tree on values rather than indices.
 	</Resource>
 </Resources>
 
+## Introduction
+Suppose you want to support the following queries:
+
+- Given a range, count the number of occurrences of value $x$.
+- Given a range, find the $k$ smallest element
+
+With a wavelet tree, you can easily support these queries in $\mathcal{O}(\log M)$ time,
+where $M$ is the maximum value in the array.
+
+### Wavelet Tree Structure
+
+To answer value-based queries efficiently, we'll create a segment tree where each node represents a range of values, instead of indices. Just like a normal segment tree, each subsequent level splits the range into two halves. Note that an index can appear in at most $\log(M)$ nodes.
+
+<Spoiler title="Wavelet Tree Visualization">
```

Comment:
why's this spoilered

Reply:
I wanted to include a popup menu so users can choose whether they want to view the visualization or continue reading.

Reply:
Should I get rid of this?

Reply:
the spoiler, yeah
the actual visualization, no